### ISSUE TITLE
Reviewing pics with a lot of files is too slow

### ISSUE THREAD

**Issue by naoliv** [mapillary/mapillary_issues#1703]
I was trying to quickly find one image from a sequence and noted that trying to review a lot of pics using Mapillary basically kills the phone (it gets really slow).

Sometimes it even crashes ("Mapillary app stopped" in English):
![](http://i.imgur.com/FVz0yMx.png)

Or, when I am able to click on some sequence, it takes a long time to have the pics on screen:

![](http://i.imgur.com/n7QMoax.png)

Unfotrunatelly it's faster to use some gallery app than Mapillary.

Mapillary version is 1.75 and the phone is running Android 6.0

**Comment by peterneubauer** [comment]
That's a problem with having large amounts of fields to deal with. At some point this takes a lot of memory, even if the app generates thumbnails. Is even the last picture (gridview) slow? What phone are you using?

**Comment by naoliv** [comment]
It's a Motorola Moto G second generation (1 GB RAM + quad-core 1.2 GHz Cortex-A7)
It's slow even if I enter one small sequence (with 6 photos, for example). Even though I am seeing a small sequence, it seems that Mapillary keeps processing all the files in background.

**Comment by iav** [comment]
Same with Samsung Galaxy Note 4 - mega power phone.
Just 900 photos in 12 sequences looks near to complete show stopper.
And seems there no any cache! Looks like thumbnails generates from the scratch on every enter into the sequence, then - on every exit from sequence into sequences list.

Other case: position in sequences list resets every time when list displlayed. It's not so easy to understand in list of sequence, whar sequence I just complete review and can set checkmark "upload".

Just scroll sequence list 2 position up now takes more than 30 seconds, and 1 try from every 2 maces app to crash.
It's near to complete unusable.

2 year ago same review was nuch easily on a much smaller devices.
Please, get thumbnail caching back, and allow a marking of a sequence as "ready to upload" inside sequence view, not only from sequence list.

**Comment by drkludge** [comment]
Another perspective is the file naming used or just the number of files. If I am out for a 12 hour run, then I will have 13,000 files.  Not only is Mapillary slow, the rest of the phone is slow. Moreover, I've had to log my photos to an SD card.  That may add to the problem.  Finally, I have an even older phone but this feels like a directory hashing issue.

**Comment by peterneubauer** [comment]
I just checked - the thumbs are generated when the main image is saved, and not recreated when you multiple times look at the sequence. You can check the thumbs cache by 

    adb shell
    ls /sdcard1/Android/data/app.mapillary/files/mapillary/.thumbs/   

And check the creation dates.

Also @drkludge 13000 files is a lot for the phone to handle, especially when we count the number of files to upload and need to iterate 13.000 items in the file system to give you an accurate estimate. It's hard to have anything that spans all images go fast at that point, but you have a point - we should probably start investigating use cases over 10000 photos and profile for them.

**Comment by iav** [comment]
Yes!
I have swiched off "generate thumbnails", because that option makes capturing very rare.
I hopes it will create thumbnails during review.

**Comment by peterneubauer** [comment]
No, not of you switched off the option. It is there since users want to
save space on the device. Do that is why your fill image will have to be
loaded and resized for every access, only using memory caching.

**Comment by naoliv** [comment]
Taking a look here and I saw that `generate thumbnails` is also off.
I don't remember if I have turned it off somewhere in the past, but this could explain why it's slow when reviewing the pictures.

**Comment by iav** [comment]
But if I switch option on - capturing is very slow on my device.
Is it possible to create cache on disk not during capture, but on first usage?

**Comment by peterneubauer** [comment]
Well,
the first view of long sequences then will be very slow, which is why the cache was created - reviewing images was slow. However, @rsahlin is working on a revamp here to hopefully generate the thumbs during the new Camera2 async loop, we will see if that is better.

**Comment by iav** [comment]
But still no way to use cache if it not created during capture?

**Comment by peterneubauer** [comment]
Well, we could have a switch for cache that says

Create eager
Create lazy
Create never

...

**Comment by iav** [comment]
Fine!

-----Original Message-----

**Comment by GITNE** [comment]
> ```$ ls -l /sdcard1/Android/data/app.mapillary/files/mapillary/.thumbs```

I do not want to sound like a smart ass but shouldn't thumbnails go into
```Java
android.content.ContextWrapper.getCacheDir()
```
or
```Java
android.content.ContextWrapper.getExternalCacheDir()
```
or
```Java
android.content.ContextWrapper.getExternalCacheDirs()
```
respectively?

IMHO thumbnails should be generated after a photo has been taken or at review in the gallery for faster lookup later. This data can be considered discardable at any time and it is probably best let the runtime handle its lifetime.

**Comment by peterneubauer** [comment]
Well,
problem is that we introduced the thumbnail caching exactly for the reason above - reviewing a sequence of several hundred photos in grid view without it is going to take a long time, and users don't want to want for lazy caching first time. The thumbnails are already generated in a background thread, we might be able to do it even smarter.

The placement of the thumbs into the mapillary dir is because users want to copy all content easily to the computer and upload/review from there. We probably can put it in the default cache also, but it's harder to find. Also, the lifetime is not discardable at any time, because recreating the thumbs is taking considerable time, and we want to delete them only after upload of the originals, not before.

**Comment by peterneubauer** [comment]
Closing this, since the only improvement we can do here is to find a better way to generate thumbs in the background, which is another issue.

**Comment by GITNE** [comment]
> problem is that we introduced the thumbnail caching exactly for the reason above - reviewing a sequence of several hundred photos in grid view without it is going to take a long time, and users don't want to want for lazy caching first time. The thumbnails are already generated in a background thread, we might be able to do it even smarter.

Understandable, and it is great that thumbnails already are created on a background thread. However, have you considered putting the original photos into [```android.os.Environment.getExternalStoragePublicDirectory(android.os.Environment.DIRECTORY_PICTURES)```](https://developer.android.com/reference/android/os/Environment.html#getExternalStoragePublicDirectory%28java.lang.String%29) or [```android.app.DownloadManager.addCompletedDownload(/*photo key etc.*/, /*photo sequnce key etc.*/, true, "image/jpeg", /*path to photo*/, /*size of photo*/, false)```](https://developer.android.com/reference/android/app/DownloadManager.html#addCompletedDownload%28java.lang.String,%20java.lang.String,%20boolean,%20java.lang.String,%20java.lang.String,%20long,%20boolean%29)? These locations are usually scanned by dedicated gallery apps which probably employ some more sophisticated thumbnail caching strategy than any strait forward strategy a common developer come up with. The later one usually just creates a link and index entry which does not cost much. Then, you would not need to bother about implementing your own gallery and thumbnail caching strategy. Plus, most gallery apps honor organization of images which should also enable you to organize photos in sequences (at least when it comes to rendering).

I am aware that there are caveats to external gallery apps, especially because you may lose control over directly managing photos. However, with [```android.app.DownloadManager.addCompletedDownload()```](https://developer.android.com/reference/android/app/DownloadManager.html#addCompletedDownload%28java.lang.String,%20java.lang.String,%20boolean,%20java.lang.String,%20java.lang.String,%20long,%20boolean%29) you would need to make the photo file world readable only while retaining control over its lifetime on the device. Another caveat is that every time the user wants to view photos from the Mapillary app the view has to jump to an external activity. But, this trade off for better caching strategy may be worth it.

> The placement of the thumbs into the mapillary dir is because users want to copy all content easily to the computer and upload/review from there. We probably can put it in the default cache also, but it's harder to find.

AFAICT [```getCacheDir()```](https://developer.android.com/reference/android/content/ContextWrapper.html#getCacheDir%28%29) returns an application specific path, something like ```/cache/app.mapillary```, if not, you can still create sub-directories in the path returned. So, things can still be neatly ordered and easy to find, instead of just getting lumped in ```/cache```. Besides, all that would go in there are just thumbnails anyway, nothing most users are really interested in (and probably unusable by most third-party software anyway), hence no need to find it *easily*, even though it would not be especially harder to find than going to ```/Android/data/app.mapillary/files/mapillary/.thumbs```.

> Also, the lifetime is not discardable at any time, because recreating the thumbs is taking considerable time, and we want to delete them only after upload of the originals, not before.

Hmm, gallery thumbnails is indeed classic cache data, not configuration or result data which really needs to be persistent. Sure, you can argue that because of the comparatively high cost that may go into creating this cache data it becomes *less* discardable. But, maybe that's the trick? Lowering the cost of creating thumbnails would make it *more* discardable again. Caching data is really about smart sharing of storage space and being a *good* *neighbor* app. Everybody would like to have unlimited persistent storage space where you do not need to cache data but this is not the reality. So, saying that you *must* keep thumbnails until the originals are off device is no argument here because it is just one rule when cache data *definitely* becomes invalid, nothing more. The problem here is bad caching strategy, or rather the lack of, not cache validity or the lifetime of cached objects.

But, I agree, some devices mount their cache on ```tmpfs``` which would discard all thumbnails on reboot.

**Comment by HolgerJeromin** [comment]
I would bet that Reviewing images in another app would be no good user experience. 

Which app has a large delete button in the middle? 

How can the gallery app transfer the information for "review is ready" , free to upload  this sequence. 

The images and the thumbs have to be in a path with much space. Splitting both path would probably results in a full phone and an empty SD card or something like that.

**Comment by GITNE** [comment]
> How can the gallery app transfer the information for "review is ready" , free to upload this sequence.

The are multiple ways one could do this.
* The user could put images ready for upload in a pre-defined "Mapillary upload" category, folder, or etc in the gallery app
* The Mapillary app can use [```android.app.Activity.startActivityForResult(android.content.Intent, int)```](https://developer.android.com/reference/android/app/Activity.html#startActivityForResult%28android.content.Intent,%20int%29) or [```android.app.Activity.startActivityForResult(android.content.Intent, int, android.os.Bundle)```](https://developer.android.com/reference/android/app/Activity.html#startActivityForResult%28android.content.Intent,%20int,%20android.os.Bundle%29) to start the gallery app for a specific result
* In any event, the Mapillary app can always ask in [```android.app.Activity.onActivityResult(int, int, android.content.Intent)```](https://developer.android.com/reference/android/app/Activity.html#onActivityResult%28int,%20int,%20android.content.Intent%29) if the user wants to upload any changes to Mapillary photos now

But, I agree, relying on an external gallery app which may not even be installed can be tricky. :unamused:

> The images and the thumbs have to be in a path with much space.

No, this is not true.

> Splitting both path would probably results in a full phone and an empty SD card or something like that.

When would this ever happen? Remember, by definition thumbnails are weakly linked **to** the original **not** the other way around.

**Comment by GITNE** [comment]
I have just thought a little bit longer about generating thumbnails. The most costly thing when generating thumbnails is reading the entire original image. Resizing is pretty fast, especially with nearest neighbor quality which should be good enough for thumbnails. This operation should take no longer than some one digit milliseconds. So, have you thought about using progressive encoding on the original photos? With progressive encoding you do not have to read the *entire* image. This could provide tolerable lazy loading performance in the gallery (even when thumbnail caching is disabled) where for most thumbnails it would probably suffice to decode just the first or second scan of the progressive encoding.

Another option might be to embed the thumbnail in the original jpeg itself. Of course, this would render any potential caching strategy employing the system cache inapplicable. But again, you would not need to read the *entire* original image, just the thumbnail. This thumbnail should be also easy discard before upload.

**Comment by GITNE** [comment]
The slowdowns @naoliv and @iav are experiencing may also come from the background thread decoding full photos or generating thumbnails not terminating quickly or soon enough after the grid view has been closed. Check this nice [video](https://www.youtube.com/watch?v=s4eAtMHU5gI) explaining this common pitfall.

**Comment by peterneubauer** [comment]
Yeah,
embedding the thumbnail in the JPEG would be good, and we are going to see
if there are image components in Android that can use that info - I didn't
find anything good there yet.
